Add support for namespaced attributes on SVG Elements

asmiller opened the pull request:
The issue with attributes on SVG elements was referenced in #21 but remains unfixed. Namespaced attributes are required for SVG icons and nodes that use xlink:href, two common use cases of SVGs on the web.

This pull request checks if the Element was created the SVG namespace (PR #26) and if so switches from setAttribute() to setAttributeNS(). It also will set "xlink:href" attributes in the proper namespace.

sparhami commented:
This doesn't really solve the problem in general. Other namespaced attributes won't work with this. Also, we don't need to use setAttributeNs for non-namespaced svg attributes.

asmiller commented:
As far as I can tell, xlink is the only attribute namespace that's required by browsers, so I think it would be ok to support it without also adding support for custom namespaces.

setAttribute is probably ok for SVG elements, but it's recommended to use setAttributeNS instead. Some browsers have issues when setAttribute is called on an element with a non-html namespace (3.4 from [http://xmlgraphics.apache.org/batik/faq.html](http://xmlgraphics.apache.org/batik/faq.html))

jridgewell commented:
I like the idea of using `:` to designate the attribute namespace: `NAMESPACE:ATTR`.

sparhami commented:
A few things:

1. Apparently XML namespaces for attributes aren't valid in HTML5. Rather than looking up the DOM for the namespace declaration, it seems things like `xlink:href` are just magic attributes that are created in the correct namespace. This seems to indicate handling arbitrary namespaces isn't really worth it.

2. The [MDN docs](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute) for SVG call out there being several attributes:

  - xlink:actuate
  - xlink:arcrole
  - xlink:href
  - xlink:role
  - xlink:show
  - xlink:title
  - xlink:type
  - xml:base
  - xml:lang
  - xml:space

  React seems to [support](https://github.com/facebook/react/blob/a261c45ccffacfbfea6808be5b6363c78a6a6bbe/src/renderers/dom/shared/SVGDOMPropertyConfig.js) all of these. Do we want to support all of these? I'm not sure anything but `xlink:href` is actually useful though.

3. We should also support removing namespaced attributes.

4. I'm not sure it is really worth checking only on SVG elements. Maybe just add an assert that is only used on SVG elements (those `xlink:href` can be used on [MathML elements](https://developer.mozilla.org/en-US/docs/Web/MathML/Attribute)).

asmiller commented:
I updated the PR to check against all of the known SVG attributes listed by @sparhami. I added a dictionary that lets us easily add other attribute namespaces if needed.

Rather than check that the element _is_ in the SVG namespace, I switched to check that the element is _not_ in the HTML namespace. This means the same method should work for MathML or other element types that have custom namespaces.

I ran some tests on the removeAttribute method ([jsfiddle](https://jsfiddle.net/2xL864qh/)). Both Chrome and Firefox can remove a namespaced attribute with the removeAttribute() method when the prefix is included in the name. But the removeAttributeNS method only works when the prefix is stripped off.
Since setAttributeNS works with names that include the namespace prefix, I'd rather set using that and remove with removeAttribute. That way we don't have to parse or split the name argument at all, since it comes in with the prefix attached.

sparhami commented:
I don't want to have the check for the XHTML namespace even. I'd rather just add an assert for it. If that means with asserts stripped, we create a `xlink:href` attribute for a div with the xlink namespace, I'm fine with that.

I'd like to also avoid using `setAttributeNS` for all `svg` content. It is a lot slower than `setAttribute`. I think if the attribute namespace is undefined, we should just use `setAttribute`. Otherwise, use `setAttributeNS`.

asmiller commented:
I'm fine with both ideas.
- In the event that an attribute named `xlink:href` on a non-svg element triggers a `setAttributeNS` call instead of `setAttribute`, the added namespace is essentially transparent to the user.
- I wasn't able to reproduce any of the supposed issues using `setAttribute` instead of `setAttributeNS` on `svg` elements.

sparhami commented on src/attributes.js:
```diff
@@ -34,7 +51,12 @@ const applyAttr = function(el, name, value) {
     el.removeAttribute(name);
   } else {
-    el.setAttribute(name, value);
+    let attrNS = attributeNSMap[name];
```
Should be `const`. Everything else looks fine to me.

jridgewell commented on src/attributes.js:
```diff
@@ -21,6 +21,23 @@ import {
+ */
+const xlinkNS = 'http://www.w3.org/1999/xlink';
+const xmlNS = 'http://www.w3.org/XML/1998/namespace';
+const attributeNSMap = {
```
We should make this prototype-less. There are some weird methods like `watch` on a default object.

asmiller replied:
Good catch, this should use createMap(). see #224
